src/apphosting/secrets/index.ts

```diff
@@ -0,0 +1,57 @@
+import { FirebaseError } from "../../error";
+import * as gcsm from "../../gcp/secretManager";
+import { FIREBASE_MANAGED } from "../../gcp/secretManager";
+import { isFunctionsManaged } from "../../gcp/secretManager";
+import * as utils from "../../utils";
+import * as prompt from "../../prompt";
+
+/**
+ * Ensures a secret exists for use with app hosting, optionally locked to a region.
+ * If a secret exists, we verify the user is not trying to change the region and verifies a secret
+ * is not being used for both functions and app hosting as their garbage collection is incompatible
+ * (client vs server-side).
+ */
+export async function upsertSecret(
+  project: string,
+  secret: string,
+  location?: string,
+): Promise<boolean> {
+  try {
+    const existing = await gcsm.getSecret(project, secret);
+    const replication = existing.replication?.userManaged;
+    if (
+      location &&
+      (replication?.replicas?.length !== 1 || replication?.replicas?.[0]?.location !== location)
+    ) {
+      utils.logLabeledError(
+        "apphosting",
+        "Secret replication policies cannot be changed after creation",
+      );
+      return false;
+    }
+    if (isFunctionsManaged(existing)) {
+      utils.logLabeledWarning(
+        "apphosting",
+        `Cloud Functions for Firebase currently manages versions of ${secret}. Continuing will disable ` +
+          "automatic deletion of old versions.",
+      );
+      const stopTracking = await prompt.confirm({
+        message: "Do you wish to continue?",
+        default: false,
+      });
+      if (!stopTracking) {
+        return false;
+      }
+      delete existing.labels[FIREBASE_MANAGED];
+      await gcsm.patchSecret(project, secret, existing.labels);
+    }
+    // TODO: consider whether we should prompt a user who has an unmanaged secret to enroll in version control.
+    // This may not be a great idea until version control is actually implemented.
+  } catch (err: any) {
```

Comment:
can we update this try catch to just wrap L20 to keep the try block focused? As it stands now, if `patchSecret` on L46 errors out for whatever reason, the user gets an unexpected error message